File: 4-median-of-two-sorted-arrays.cpp
/*
  Source : https://leetcode.com/problems/median-of-two-sorted-arrays/
  Author : Alex Xu
  Date   : June 10, 2015
  Problem:
  There are two sorted arrays nums1 and nums2 of size m and n respectively. Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
*/

#inlcude <algorithm>

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int total = nums1.size() + nums2.size();
        
        if (total % 2 == 1) { // odd number {1,2,3} -> 2
            return findKthSmallest(nums1, 0, nums1.size(), nums2, 0, nums2.size(), total/2 + 1);
        } else { // even number {1, 2, 3, 4} -> (2+3)/2.0
            return (findKthSmallest(nums1, 0, nums1.size(), nums2, 0, nums2.size(), total/2) + 
                    findKthSmallest(nums1, 0, nums1.size(), nums2, 0, nums2.size(), total/2 + 1)) / 2.0;
        }
    }
    
    int findKthSmallest(vector<int>& A, int lowA, int lenA, vector<int>& B, int lowB, int lenB, int k) {
        // always assume nums1.size() < nums2.size()
        if (lenA > lenB)
            return findKthSmallest(B, lowB, lenB, A, lowA, lenA, k);
        
        // exit condition: m is empty or k==1    
        if (lenA == 0)
            return B[k-1];
        if (k == 1)
            return std::min(A[lowA], B[lowB]);
        
        // devide k into two parts, kA in A and kB in B, where (kA + Kb = k)
        int kA = std::min(k/2, int(A.size()));
        int kB = k - kA;
        
        if (A[lowA + kA-1] < B[lowB+kB-1]) // ignore A[lowA,...lowA+lenA]
            return findKthSmallest(A, lowA + kA, lenA - kA, B, lowB, lenB, k-kA);
        if (A[lowA+kA-1] > B[lowB+kB-1]) // ignore B[lowB,...lowB+lenB]
            return findKthSmallest(A, lowA, lenA, B, lowB + kB, lenB - kB, k-kB);
        else 
            return A[lowA+kA-1];
    }
    
private:
    double findMedianInSortedArray(vector<int>& v) {
        int n = v.size();
        if (n == 0)
            return 0;
        
        if (n % 2 == 0)
            return (v[n/2 - 1] + v[n/2]) / 2.0;
        else
            return v[n/2];
    }
};

File: 4-median-of-two-sorted-arrays.py
##
# Source : https://leetcode.com/problems/median-of-two-sorted-arrays/
# Author : Alex Xu
# Date   : Feb 20, 2018
# Problem:
# There are two sorted arrays nums1 and nums2 of size m and n respectively.
# Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
#
# Your runtime beats 98.63 % of python3 submissions.
##
class Solution:
        def findMedianSortedArrays(self, nums1, nums2):
            """
            :type nums1: List[int]
            :type nums2: List[int]
            :rtype: float
            """
            total = len(nums1) + len(nums2)
            if total % 2 == 1:
                # odd number [1, 2, 3, 4, 5] -> list[5/2] = list[2] = 3: index 2 is the 3rd
                return self.findKthSmallest(nums1, nums2, total//2 + 1)
            else:
                # even number [1, 2, 3, 4] -> (list[4/2] -1 + list[4/2])/2.0 = (list[1] + list[2])/2.0, index 1 is the 2nd and index 2 is the 3rd
                return (self.findKthSmallest(nums1, nums2, total//2) + self.findKthSmallest(nums1, nums2, total//2 + 1)) / 2.0
            
        def findKthSmallest(self, a, b, k):
            """
            :type a: list[int]
            :type b: list[int]
            :type k: int
            """
            if len(a) > len(b):
                return self.findKthSmallest(b, a, k)
            
            # exit condition: a is empty or k==1
            if not a:
                return b[k-1]
            if k == 1:
                return min(a[0], b[0])
            
            # devide k into two parts ka elements in a and kb elements in b, where k = ka + kb
            ka = min(k//2, len(a))
            kb = k - ka
            
            if a[ka - 1] < b[kb - 1]:
                return self.findKthSmallest(a[ka:], b[:kb], kb) # ignore a[:ka]
            elif a[ka -1] > b[kb - 1]:
                return self.findKthSmallest(a[:ka], b[kb:], ka) # ignore a[:kb]
            else:
                return a[ka-1]; # or return b[k-1]
